text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+other instances of the set, and the `SharingError` type allows modelling potentially-sharing stores, such as a small
+store which cannot be shared if its handles currently point to inline memory.
+
+
+##  TypedHandle
+
+A straightforward extension is to define a `TypedHandle<T, H>` type, which wraps a handle (`H`) to a block of memory
+suitable for an instance of `T`, and also wraps the metadata of `T`.
+
+The `Store` methods can then be mimicked, with the additional type information:
+
+-   `resolve` returns an appropriate pointer type, complete with metadata.
+-   Layouts are computed automatically based on the type and metadata.
+-   Growing and Shrinking on slices take the target number of elements, rather than a more complex layout.
+
+And because `resolve` and `resolve_mut` can return references -- being typed -- they can borrow the `store` (shared) to
+ensure it's not moved nor dropped.
+
+
+##  UniqueHandle
+
+A further extension is to define a `UniqueHandle<T, H>` type, which adds ownership & borrow-checking over a
+`TypedHandle<T, H>`.
+
+That is:
+
+```rust
+impl<T: ?Sized, H: Copy> UniqueHandle<T, H> {
+    pub unsafe fn resolve<'a, S>(&'a self, store: &'a S) -> &'a T
+    where
+        S: Store;
+
+    pub unsafe fn resolve_mut<'a, S>(&'a mut self, store: &'a S) -> &'a mut T
+    where
+        S: Store;
+}
+```
+
+Those two methods are `unsafe` as a dangling handle cannot be soundly resolved, and a valid handle may not necessarily
+be associated with a block of memory containing a valid instance of `T` -- it may never have been constructed, it may
+have been moved or dropped, etc...
+
+On the other hand, those two methods guarantee:
+
+-   Proper typing: if the handle is valid, and a value exists, then that value is of type `T`.
+-   Proper borrow-checking: the handle is the unique entry point to the instance of `T`, hence the name.
+-   Proper pinning: even if the store does not implement `PinningStore`, borrowing it ensures that it cannot be moved
+    nor dropped. If the store implements `StableStore`, this means that the result of `resolve` and `resolve_mut` can be
+    used without fear of invalidation.
+
+And that's pretty good, given how straightforward the code is.
+
+
+##  Compact Vectors
+
+How far should DRY go?
+
+One limitation of `Vec<u8, InlineStore<[u8; 16]>>` is that it contains 2 `usize` for the length and capacity
+respectively, which is rather unfortunate.
+
+There are potential solutions to the issue, using separate traits for those values so they can be stored in more compact
+ways or even elided in the case of fixed capacity.
+
+The `Store` API could be augmented with a new marker trait with associated constants / types describing the limits of
+what it can offer such as minimum/maximum capacity, to support automatically selecting (or constraint-checking) the
+appropriate types.
+
+Since those extra capabilities can be brought in by user traits for now, I would favor adopting a wait-and-see approach
+here.
```

Comment:
It's not exactly trivial to add post facto, since all storages would need to provide the "potentially limited storage" API for it to be meaningfully useful. But combined with pattern refined types, I think adding to `Store` something like:

```rust
pub trait Store {
    const MAX_STORAGE: usize = isize::MAX as usize;
    const MIN_STORAGE: usize = 0;
}
```

should get the job done, since you can have e.g.

```rust
struct Vec<T, S> {
    handle: S::Handle,
    len: usize @ 0..={S::MAX_STORAGE / size_of::<T>()},
    cap: usize @ {S::MIN_STORAGE / size_of::<T>()}..={S::MAX_STORAGE / size_of::<T>()},
    store: S,
    marker: PhantomData<Box<[T]>>,
}
```

and that should hopefully theoretically be able to at least optimize for single-valued `cap` and `cap`/`len` which are always less than `uNN`... though also maybe not, depending on coercions (e.g. `(&usize @ PAT) as &usize`); pattern restricted types might just give you niches, not size compression. Interim solutions that use other solutions should definitely be left to 3rd party, since that type jujitzu is *definitely* beyond anything else std has exposed. (The `Store` trait hierarchy is already close to if not the most involved API in std, even after the massive simplifications v3 gives over v2 and v2 gives over v1.)

Reply:
I'm still not convinced that this would the right point of customization, to be honest.

For example, I regularly use `u32` for length (and `u8` for capacity, as the exponent of a power-of-2) even with the `Allocator` API not because the `Allocator` cannot provide more, but because I don't care about supporting more.

In fact, my implementation of `InlineVec` uses `u16` regardless, since I have uses for more than 255, but not use over 64K elements.

Requiring to wrap the store into an adapter that limits the sizes every time you wish to save up some bytes seem more complicated than directly plugging the types you want to use in the first place.